MetaDrive Vectorized Encoding Part VI: Implement the BEV observation

breakds opened the pull request:
# Motivation

To make a better interface for the coexistence of both the vectorized and raster encoding for MetaDrive, the design is to use `create_environment`'s `env_name` to determine whether we want the vectorized version or the raster version.

i.e.

```
alf.config(
    'create_environment', env_name='Vectorized', num_parallel_environments=36)
```
or
```
alf.config(
    'create_environment', env_name='BirdEye', num_parallel_environments=36)
```

The two will spawn the same environment wrapper `AlfMetaDriveWrapper` (as defined in `suite_metadrive.py`), but with different underlying environment, where they differ mostly on having different `Observation` instance. We already have the `VectorizedObservation`, and this PR adds its raster (BEV) counter part, called the `BirdEyeObservation`.

# Solution

Implement `BirdEyeObservation`. It wraps MetaDrive's BEV observation, with an extra output on velocity history. It is expected that the model will combine the velocity into the BEV.

# Testing

Together with the next PR, verified that the training is still working for BEV + CNN. Also verified the output BEV with visualization to make sure it is the same as before.

breakds committed:
Implement the BEV observation

breakds commented on alf/environments/metadrive/environments.py:
```diff
@@ -35,6 +35,34 @@ class VectorizedTopDownEnv(metadrive.MetaDriveEnv):
+        return self.get_single_observation().observation_spec
+
+
+class BirdEyeTopDownEnv(metadrive.MetaDriveEnv):
```
The diff is presented in an odd way. Actually the code starting from here to the end of the file is the newly added code. The reason it is presented in such a way is that I removed unrelated options from the `config.update({})` call above.

breakds committed:
Fix image channel first

Haichao-Zhang commented on alf/environments/metadrive/sensors.py:
```diff
@@ -167,3 +169,85 @@ def reset(self, env, vehicle=None):
+            env_config: MetaDrive's environment configuration.
+            velocity_steps: The number of historical steps for the velocity.
+            velocity_normalization: The velocities (in m/s) will be normalizaed
+                by this factor before producing the feature.
```
normalizaed

breakds replied:
Fixed.

Haichao-Zhang commented on alf/environments/metadrive/sensors.py:
```diff
@@ -167,3 +169,85 @@ def reset(self, env, vehicle=None):
+        """
+        img = super().observe(vehicle)
+
+        self._velocity_history[:-1] = self._velocity_history[1:]
```
will it be useful to include a mask for indicating the padding for velocity_history?

Haichao-Zhang replied:
thought about this again and the mask might be unnecessary in this case since the initial velocities can be safely assumed to be zero.

breakds committed:
Fix typo

breakds commented:
Addressed comments.

Haichao-Zhang commented:
> Addressed comments.

Need rebase. 
The next few PRs in this series are likely to have the same problem after one of them is merged.

breakds commented:
> > Addressed comments.
> 
> Need rebase. The next few PRs in this series are likely to have the same problem after one of them is merged.

Thanks. Actually it was rebased in a wrong way. It should be fixed now.

Haichao-Zhang commented:
> > > Addressed comments.
> > 
> > 
> > Need rebase. The next few PRs in this series are likely to have the same problem after one of them is merged.
> 
> Thanks. Actually it was rebased in a wrong way. It should be fixed now.

There is a new commit to the master fixing the broken tests.

breakds commented:
Rebased and re-tested. PTAL. Thanks!
